### DISCUSSION TITLE
Does new coffea with DaskExecutor function with BaseSchema?

### DISCUSSION THREAD

**Discussion by tomeichlersmith** [scikit-hep/coffea#937] [Q&A] [answered]
I ask this seemingly absurd question because I noticed (with the runner configuration below), that my processor does not even get to the `process` call before a `dask` error is thrown.
```python
    runner = processor.Runner(
        executor = processor.DaskExecutor(
            client = client,
            compression = None,
            retries=0,
            worker_affinity = True
        ),
        pre_executor = processor.IterativeExecutor(),
        schema = BaseSchema,
        processor_compression = None,
        metadata_cache = {},
    )
```

### Dask Error
I can copy in the full error, but the rest seems to be from the other workers being killed.
```
<dask.highlevelgraph.HighLevelGraph object at 0x7f18dec657e0>                                                         
 0. num-9649a8c34eb4f5de4ee14279d985e6d5                                                                              
>.                                                                                                                    
Traceback (most recent call last):                                                                                    
  File "/export/scratch/users/eichl008/ldmx/eat/v14/analysis/coffea/.local/lib/python3.10/site-packages/distributed/protocol/pickle.py", line 63, in dumps
    result = pickle.dumps(x, **dump_kwargs)
AttributeError: Can't pickle local object 'NanoEventsFactory.events.<locals>.<lambda>'
```

Am I mis-configuring the `DaskExecutor`? Do I need to use a different schema in order to avoid the local objects produced by the `NanoEventsFactory`?

**Comment by lgray** [comment]
Are you doing this in coffea 2023?

**Reply by lgray** [reply]
Ah, yes it is.

**Reply by tomeichlersmith** [reply]
Yes

```
coffea==2023.10.0rc1
```

<details>
  <summary>pip freeze</summary>

```
awkward==2.5.0                                                                                                                                                                                                                                
awkward-cpp==26                                                                                                                                                                                                                               
awkward0==0.15.5  
boost-histogram==1.4.0
cachetools==5.3.2
click==8.1.7
cloudpickle==3.0.0
coffea==2023.10.0rc1
contourpy==1.2.0
correctionlib==2.4.0
cycler==0.12.1
dask==2023.11.0
dask-awkward==2023.11.3
dask-histogram==2023.10.0
distributed==2023.11.0
fonttools==4.44.3
fsspec==2023.10.0
hist==2.7.2
histoprint==2.4.0
importlib-metadata==6.8.0
Jinja2==3.1.2
kiwisolver==1.4.5
llvmlite==0.41.1
locket==1.0.0
lz4==4.3.2
markdown-it-py==3.0.0
MarkupSafe==2.1.3
matplotlib==3.8.1
mdurl==0.1.2
memory-profiler==0.61.0
mplhep==0.3.31
mplhep-data==0.0.3
msgpack==1.0.7
numba==0.58.1
numpy==1.23.5
packaging==23.2
pandas==2.1.3
partd==1.4.1
Pillow==10.1.0
psutil==5.9.6
pyarrow==14.0.1
pydantic==1.10.13
Pygments==2.16.1
pyparsing==3.1.1
python-dateutil==2.8.2
pytz==2023.3.post1
PyYAML==6.0.1
rich==13.7.0
scipy==1.11.3
six==1.16.0
sortedcontainers==2.4.0
tblib==3.0.0
toml==0.10.2
toolz==0.12.0
tornado==6.3.3
tqdm==4.66.1
typing_extensions==4.8.0
tzdata==2023.3
uhi==0.4.0
uproot==5.1.2
uproot3==3.14.4
uproot3-methods==0.10.1
urllib3==2.1.0
zict==3.0.0
zipp==3.17.0
```
</details>

**Comment by lgray** [comment, marked as answer]
The Runner interface is being made obsolete, largely by the adoption of dask-awkward.

What you should do instead is:
```python3
events = NanoEventsFactory.from_root(your_files, options, ....)

proc = YourProcessor(*args)
out_collections = proc(events)
result = dask.compute(out_collections)
```

We are making things over in https://github.com/CoffeaTeam/coffea/pull/882 to deal better with the concept of datasets, you can merge that as well and use those interfaces if you like.

**Reply by lgray** [reply]
If you need to talk to a batch system then you should make the appropriate Cluster and `Client (from dask_jobqueue and distributed, respectively). You can see an example for the LPC here: https://github.com/CoffeaTeam/lpcjobqueue/blob/main/simple_example.py#L49-L52

All you'd have to do is create that client before you start making the dask taskgraph to get it all to work.

**Reply by lgray** [reply]
@tomeichlersmith did that get you sorted?

**Reply by tomeichlersmith** [reply]
yes! Apologies, my university network went down - thank you for checking in!

**Reply by tomeichlersmith** [reply]
# For Future Readers

I followed @lgray 's advice and dropped usage of the executor/runner from coffea. A slightly more fleshed out example (closer to what I actually do) is below. I haven't tested this in full but it may be a helpful starting point.

```python

p = MyProcessor(*config_args)
out = {}
for sample, data in dataset.items():
    events = NanoEventsFactory.from_root(
        { fp : 'tree_name' for fp in data['files'] },
        schema = BaseSchema,
        metadata = data['metadata'],
        permit_dask = True, # this option has been renamed on the HEAD of master but was this name in the 2023.10.0rc1 release
    ).events()
    out[sample] = p.process(events)

# print task graph or just the list of delayed calculations
print(out)
# confirm to run
input('Press Enter to continue...')
out = dask.compute(out)
```

**Reply by green-cabbage** [reply]
Hello; would you be willing to share the rest of the code? I am wondering where you specify the number of workers and/or link the process with Dask Gateway if possible. Thank you in advance!

**Reply by tomeichlersmith** [reply]
The full code is in a private repo for my experiment's analysis code, but I can share with you the function that surrounds the part I copied above.

Edit: I was looking at different beam energies here and so sometimes the "category"/"sample" name of a dataset is referred to as `beam`.

```python
# construct a dataset,
#   category -> file listing dictionary
#   the bkgd_unbiased_list just returns a python list of filepath strs
dataset = {
    beam: { 'files': bkgd_unbiased_list(beam), 'metadata': {'beam': int(beam[0])*1000} }
    for beam in ['4gev','8gev']
}

# just a callback to limit the number of files to a certain number if I want to run a shorter test
if nfilelimit is not None:
    for beam in dataset:
        dataset[beam]['files'] = dataset[beam]['files'][:nfilelimit]

# setting up of the dask client
# ncores is passed as an argument to this function
# I do not use a fully distributed dask solution because the data I'm reading
# is residing on a simple NFS server which would lock up if too many cores are reading from it
from dask.distributed import Client
client = Client(
    dashboard_address=':8888',  # start distributed scheduler locally
    threads_per_worker=1,
    n_workers=ncores,
    memory_limit='auto'
)

# create the processor and run it over each entry in the dataset
p = MyProcessor()
out = {}
for category, data in dataset.items():
    events = NanoEventsFactory.from_root(
        { fp : 'tree_name' for fp in data['files'] },
        schemaclass = BaseSchema, 
        metadata = data['metadata'],
        chunks_per_file = 1, # see comments about NFS server, just doing full files at a time to reduce disk calls
        permit_dask = True # this option has been renamed on the HEAD of master but was this name in the 2023.10.0rc1 release

    ).events()
    out[category] = p.process(events)

# print here to inspect task graph or calculations before actually running
print(out)
# and pause waiting for user input
input('compute?')
# print link to dashboard so its easier to open it after starting
print(client.dashboard_link)

# this is the call that actually runs the tasks
out = dask.compute(out)

# re-print result
print(out)

# write the resulting accumulated histograms into a pickle file for later plotting
with open(output_name, 'wb') as outf:
    pickle.dump(out, outf)
```

**Reply by green-cabbage** [reply]
Thank you for the quick reply. So I see that you manually defined a dask client with set number of workers. But where do you link the client to the coffea event and processing? Does coffea automatically detect it now? If so, is there a way to manually link your client for the coffea backend to use it?

**Reply by lgray** [reply]
Detection of the `Client` is done automatically through dask.

If you want to use a specific client you should use a context manager:
```python3
with Client(<arguments>) as client:
    events = NanoEventsFactory.from_root(...)
    out, = dask.compute(analysis(events))
```

and that will ensure the client you specify in the with block is used for execution.

**Reply by lgray** [reply]
In coffea 2023 and onwards the definition of the analysis is separate from its execution, so there's so "connection" between nanoevents and creating the definition of your analysis and the Client at all.

**Reply by green-cabbage** [reply]
I see, does that mean that we define the chunk_size and maxchunks in metadata? Suppose that I want to set chunk_size = 100 and maxchunks = 2, I should have the fileset.json file as (for e.g):
`
"/DYJetsToLL_M-50_TuneCP5_13TeV-amcatnloFXFX-pythia8/RunIISummer20UL18NanoAODv9-106X_upgrade2018_realistic_v16_L1v1-v2/NANOAODSIM": {
"files": {
      "root://grid-cms-xroo....20E.root": {
        "object_path": "Events",
        "steps": [
          [
            0,
            100
          ],
          [
            100,
            200
          ],
          ],
          "uuid": <some id>
          }
`

Does this sound about right?

**Reply by lgray** [reply]
Use `dataset_tools.preprocess` and `dataset_tools.max_chunks`. That'll get you what you want.

The tests are actually pretty readable so just have a look there. Docs will come :-)

**Reply by lgray** [reply]
Actually, we now have some example based documentation for this. At least in terms of use in some simple but realistic examples. https://github.com/CoffeaTeam/coffea/blob/master/binder/processor.ipynb
